### PR TITLE
Light bulb

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
WIP
"Resolve open namespaces" works in both 2013 and 2015.
Please, review.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
Two tests fail: time outed DepthTagger one (as it does often) and a formatting one:

```
2) Test Failure : FSharpVSPowerTools.Tests.FormattingCommandTests.should be able to format document
     Expected: "
type Type = 
    | TyLam of Type * Type
    | TyVar of string
    | TyCon of string * Type list
    override this.ToString() = 
        match this with
        | TyLam(t1, t2) -> sprintf "(%s -> %s)" (t1.ToString()) (t2.ToString())
        | TyVar a -> a
        | TyCon(s, ts) -> s
"
Actual: "
 
type Type
    = TyLam of Type * Type
    | TyVar of string
    | TyCon of string * Type list
    with override this.ToString () =
            match this with
            | TyLam (t1, t2) -> sprintf "(%s -> %s)" (t1.ToString()) (t2.ToString())
            | TyVar a -> a
            | TyCon (s, ts) -> s
"
  Expected string length 284 but was 305. Strings differ at index 1.
  Expected: "\ntype Type = \n    | TyLam of Type * Type\n    | TyVar of strin..."
  But was:  "\n\ntype Type\n    = TyLam of Type * Type\n    | TyVar of string\n..."
  -------------^
 
at FSharpVSPowerTools.Tests.FormattingCommandTests.should be able to format document() in C:\projects\visualfsharppowertools\tests\FSharpVSPowerTools.Tests\FormattingCommandTests.fs:line 53
```
I could not run tests from `FSharpVSPowerTools.Tests` locally for months because:
```
Could not load file or assembly 'Microsoft.VisualStudio.Text.Logic, Version=11.0.0.0, Culture=neutra
l, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
```

File: paket.lock [hunk 1]
```diff
@@ -2,15 +2,15 @@ NUGET
   remote: http://nuget.org/api/v2
   specs:
     Expression.Blend.Sdk (1.0.2)
-    FAKE (3.36.0)
-    Fantomas (1.8.0-beta)
-      FSharp.Compiler.Service (>= 1.4.0-beta)
+    FAKE (4.0.3)
+    Fantomas (1.7.0)
```
Comment (dungpa) [inline paket.lock]: The formatting test can be fixed by reverting to Fantomas 1.8.0-beta.

The other tests fail due to AppVeyor timing changed (they probably moved us to a slow image).

File: src/FSharpVSPowerTools/FSharpVSPowerTools.csproj [hunk 2]
```diff
@@ -218,6 +218,12 @@
       <IncludeOutputGroupsInVSIX>BuiltProjectOutputGroup%3bBuiltProjectOutputGroupDependencies%3bGetCopyToOutputDirectoryItems%3bSatelliteDllsProjectOutputGroup%3b</IncludeOutputGroupsInVSIX>
       <IncludeOutputGroupsInVSIXLocalOnly>DebugSymbolsProjectOutputGroup%3b</IncludeOutputGroupsInVSIXLocalOnly>
     </ProjectReference>
+    <ProjectReference Include="..\FSharpVSPowerTools.Logic.VS2015\FSharpVSPowerTools.Logic.VS2015.fsproj">
+      <Project>{bb0c6f01-5e57-4575-a498-5de850d9fa6c}</Project>
+      <Name>FSharpVSPowerTools.Logic.VS2015</Name>
+      <IncludeOutputGroupsInVSIX>BuiltProjectOutputGroup%3bBuiltProjectOutputGroupDependencies%3bGetCopyToOutputDirectoryItems%3bSatelliteDllsProjectOutputGroup%3b</IncludeOutputGroupsInVSIX>
+      <IncludeOutputGroupsInVSIXLocalOnly>DebugSymbolsProjectOutputGroup%3b</IncludeOutputGroupsInVSIXLocalOnly>
```
Comment (dungpa) [inline src/FSharpVSPowerTools/FSharpVSPowerTools.csproj]: What do these lines mean to do?

File: src/FSharpVSPowerTools.Logic/ResolveUnopenedNamespaceTagger.fs [hunk 3]
```diff
@@ -14,10 +14,18 @@ open FSharpVSPowerTools.ProjectSystem
 open Microsoft.FSharp.Compiler
 open System.Threading
 
+[<NoComparison; NoEquality>]
+type Suggestion =
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Logic/ResolveUnopenedNamespaceTagger.fs]: Can we change this record type to an interface and use object expressions for instantiation? I think it's recommended in F# Component Design guidelines.

File: src/FSharpVSPowerTools/FSharpVSPowerTools.csproj (hunk 2 above)
Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools/FSharpVSPowerTools.csproj]: No idea.

File: paket.lock (hunk 1 above)
Comment (vasily-kirichenko) [inline paket.lock]: Done.

File: src/FSharpVSPowerTools.Logic/ResolveUnopenedNamespaceTagger.fs (hunk 3 above)
Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic/ResolveUnopenedNamespaceTagger.fs]: Done.

File: src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    |> List.map (fun s ->
+                         { new ISuggestedAction with
+                               member __.DisplayText = s.Text
+                               member __.Dispose() = ()
+                               member __.GetActionSetsAsync _ct = Task.FromResult <| seq []
+                               member __.GetPreviewAsync _ct = Task.FromResult null
+                               member __.HasActionSets = false
+                               member __.HasPreview = false
+                               member __.IconAutomationText = null
+                               member __.IconMoniker = Unchecked.defaultof<_>
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: Could you try to recover icons for VS2015 using `KnownMonikers.AddNamespace`?

References:

http://glyphlist.azurewebsites.net/knownmonikers/

https://github.com/agocke/pretend-roslyn/blob/031636f188e8c285230237c382494fc57ca40524/open/src/VisualStudio/Core/SolutionExplorerShim/DiagnosticItem/DiagnosticItem.cs#L118-L129

File: src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    [<Import; DefaultValue>]
+    val mutable ProjectFactory: ProjectFactory
+
+    interface ISuggestedActionsSourceProvider with
+        member x.CreateSuggestedActionsSource(textView: ITextView, buffer: ITextBuffer): ISuggestedActionsSource = 
+            if textView.TextBuffer <> buffer then null
+            else
+                let generalOptions = Setting.getGeneralOptions x.ServiceProvider
+                if obj.ReferenceEquals (generalOptions, null) || not generalOptions.ResolveUnopenedNamespacesEnabled then null
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: Is this effectively `===` in `Utils.fs`?

**Comment by dungpa** [comment]
Good idea on converting Resolving open namespaces first. It's the most comprehensive feature of the four. 

Would you like to convert other three features as part of this PR?

File: src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs (hunk 5 above)
Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: No, it's `LanguagePrimitives.PhysicalEquality`. I added `==` as an alias for `ReferenceEquals`.

**Comment by vasily-kirichenko** [comment]
I done Resolve unopened namespaces first in order to see how it's going. It seems works OK, so the PR can be merged. I (or someone else) will do the same for other smart taggers later.

File: src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs (hunk 4 above)
Comment (dungpa) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: @vasily-kirichenko Do you think we can do this before merging the PR?

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: I tried to use that `ImageMoniker`:

image

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: OK, there're two `ImageMoniker` structs, one is in `Microsoft.VisualStudio.ImageCatalog.dll` and another in `Microsoft.VisualStudio.Language.Intellisense.dll`, and we reference both.

Comment (dungpa) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: Do we need to reference both? I assume using full names would work?

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: It's also defined in `Microsoft.VisualStudio.Imaging.dll`

What do you mean by full names?

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: I removed references to Imaging and ImageCatalog, now the struct has no fields:

image

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: image

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: I add a commit with uncompiled code which trying to set `ImageMoniker`

Comment (dungpa) [inline src/FSharpVSPowerTools.Logic.VS2015/ResolveUnopenedNamespaceSuggestedActionsSourceProvider.fs]: These compile:

ImageMoniker(Guid=Guid "{ae27a6b0-e345-4288-96df-5eaf394ee369}", Id=90)

ImageMoniker(Guid=KnownMonikers.AddNamespace.Guid, Id=KnownMonikers.AddNamespace.Id)

(I have to add *.ImageCatalog.dll and `open Microsoft.VisualStudio.Imaging` for the second one to work.)

**Comment by vasily-kirichenko** [comment]
`ImageMoniker(Guid=Guid "{ae27a6b0-e345-4288-96df-5eaf394ee369}", Id=90)` compiles (committed), but the icon does not appear.

**Comment by dungpa** [comment]
Where do these values come from?

**Comment by vasily-kirichenko** [comment]
They are from `Microsoft.VisualStudio.ImageCatalog.dll`:

```csharp
public static ImageMoniker AddNamespace
    {
      get
      {
        return new ImageMoniker()
        {
          Guid = KnownImageIds.ImageCatalogGuid,
          Id = 90
        };
      }
    }

public static class KnownImageIds
  {
    public static readonly Guid ImageCatalogGuid = new Guid("{ae27a6b0-e345-4288-96df-5eaf394ee369}");
...
```

**Comment by vasily-kirichenko** [comment]
It seems Roslyn does not support icons either http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Suggestions/SuggestedAction.cs,241 Maybe it's just not yet implemented in VS?

**Comment by dungpa** [comment]
Darn. Let's not waste time on this anymore.

**Pull request merged**